---
title: "BONCAT Jan 2020 Decontam"
output: html_notebook
---

# Introduction
I would like to use the 'decontam' packages to remove sequences that are likely to be contamination. Many of the early seq runs in this project did not have QC data in the format of 16S copies from qPCR. Instead, I will just employ a method that looks at controls vs. samples


# Setting up environment
```{r}
library(decontam); packageVersion("decontam")
library(phyloseq); packageVersion("phyloseq")
library(tidyverse); packageVersion("tidyverse")
```

# The data
Load in all components created in the DADA2 analysis and the sample sheet
```{r}
samdf <- read_csv("../data/samdf_JAN2020.csv")
rownames(samdf) <- samdf$SAMPLE_NAME #because R - dplyr will not like making rownames, but it's fine for now.
# seqtab is the sample:ASV table made in DADA2 - it should contain all samples and ASVs
seqtab <- readRDS("../data/DADA2/seqtab_JAN2020.rds")
# Table with all ASVs and assigned taxonomy using the SILVA database (Greengenes and RDP also available)
taxtab <- readRDS("../data/DADA2/tax_silva_JAN2020.rds")
#taxtab <- readRDS("../data/DADA2/tax_DECIPHER_JAN2020.rds") This one just too many code breaks for now
# Phylogenetic tree made using DECIPHER and Phangorn - import object and pull tree from the fitGTR
fitGTR <- readRDS("../data/DADA2/fitGTR_JAN2020.rds")
```

```{r}
ps <- phyloseq(otu_table(seqtab, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxtab),
               phy_tree(fitGTR$tree)
)
ps

ps_data <- data.frame(sample_data(ps))

```

# Raw dataset stats

I will use the ampvis2 package to output some simple stats about our dataset.
```{r}
library(ampvis2)
#Combine OTU abundance table and taxonomy table from the phyloseq object "my_phyloseq_object":
obj <- ps
# Make short names for OTUs - facilitates later plotting ease when amp_heatmap has the option tax_empty = "best" (for some reason)
taxa_names(obj) <- paste0("ASV", seq(ntaxa(obj)))
# Fix OTU table layout for exporting. taxa_as_rows = FALSE was not working.
Totu_table =t(otu_table(obj))
otu_table(obj)=Totu_table
#export OTU table from phyloseq object
otutable <- data.frame(OTU = rownames(phyloseq::otu_table(obj)@.Data),
                       phyloseq::otu_table(obj)@.Data,
                       phyloseq::tax_table(obj)@.Data,
                       check.names = FALSE
                       )
otutable <- otutable %>% select(-OTU, 
                                -Genus_Number,
                                )

#Extract metadata from the phyloseq object:
metadata <- data.frame(phyloseq::sample_data(obj), 
                       check.names = FALSE
                       )
metadata <- rownames_to_column(metadata, var = "SAMPLE_ID")

# Extract phylogenetic tree from phyloseq object:
phytree <- phyloseq::phy_tree(obj)

#Load the data with amp_load:
ampvis <- amp_load(otutable, metadata,tree = phytree)
ampvis
```


Letâ€™s take a quick first look at the library sizes (i.e. the number of reads) in each sample, as a function of whether that sample was a true positive sample or a negative control:
```{r}
df <- as.data.frame(sample_data(ps)) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(ps)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=SAMPLE_OR_CONTROL)) + geom_point()
```
The library sizes of the positive samples primarily fall from 20,000 to 80,000 reads. The negative control samples have fewer reads as expected. Running decontam prevalence method with default options.

# Decontam - either prevalence or frequency
```{r}
sample_data(ps)$is.neg <- sample_data(ps)$SAMPLE_TYPE == "Control"
contamdf.prev.freq <- isContaminant(ps, method="either",  neg="is.neg", conc="MOL_PER_MIC", threshold = 0.5)
head(contamdf.prev.freq)
```

Some very low abundance taxa are identified by the frequency method.
```{r}
tax_ps <- as.data.frame(taxtab)
tax_ps <- rownames_to_column(tax_ps, "ASV")

TRUEcontamdf_prev_freq <- contamdf.prev.freq %>%
  rownames_to_column("ASV") %>%
  filter(contaminant == TRUE) %>%
  inner_join(tax_ps, by = "ASV")
```

Let's stick on the conservative side and let abundance filtering take care of the rest.

```{r}
ps.noncontam <- prune_taxa(!contamdf.prev.freq$contaminant, ps)
ps.noncontam
```

```{r}
saveRDS(ps.noncontam, "../data/decontam/ps-decontam_prev5.RData")
```

I will use the ampvis2 package to output some simple stats about our filtered dataset.
```{r}
library(ampvis2)
#Combine OTU abundance table and taxonomy table from the phyloseq object "my_phyloseq_object":
obj <- ps.noncontam
# Make short names for OTUs - facilitates later plotting ease when amp_heatmap has the option tax_empty = "best" (for some reason)
taxa_names(obj) <- paste0("ASV", seq(ntaxa(obj)))
# Fix OTU table layout for exporting. taxa_as_rows = FALSE was not working.
Totu_table =t(otu_table(obj))
otu_table(obj)=Totu_table
#export OTU table from phyloseq object
otutable <- data.frame(OTU = rownames(phyloseq::otu_table(obj)@.Data),
                       phyloseq::otu_table(obj)@.Data,
                       phyloseq::tax_table(obj)@.Data,
                       check.names = FALSE
                       )
otutable <- otutable %>% select(-OTU, 
                                -Genus_Number,
                                )

#Extract metadata from the phyloseq object:
metadata <- data.frame(phyloseq::sample_data(obj), 
                       check.names = FALSE
                       )
metadata <- rownames_to_column(metadata, var = "SAMPLE_ID")

# Extract phylogenetic tree from phyloseq object:
phytree <- phyloseq::phy_tree(obj)

#Load the data with amp_load:
ampvis <- amp_load(otutable, metadata,tree = phytree)
ampvis
```

```{r}
sessionInfo()
```

